### PR TITLE
Improve performance of COUNT (distinct x) for dictionary columns #258

### LINKED ISSUE #258: Improve performance of COUNT (distinct x) for dictionary columns
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I have large amounts of low cardinality string data (for example, 200 M rows, but only 20 distinct values). DictionaryArrays are very good for such data as they are space efficient.  

https://github.com/apache/arrow-datafusion/pull/256 adds basic query support for distinct dictionary columns but it is not a very computationally efficient imlementation. It effectively unpacks the (likely mostly deduplicated) dictionary's values row by row into a hash set to deduplicate it again. That is a lot of extra hashing work.

**Describe the solution you'd like**
It would likely be much more efficient (especially for arrays that have a small number of distinct values in their dictionary) to look at the values from the dictionary directly, first checking that each entry in the dictionary was actually used.

### PR DESCRIPTION
**Pull request by jaylmiller**
# Which issue does this PR close?
Closes #258.

# Rationale for this change
The count distinct physical expr was doing alot of unnecessary hashing when it is ran on dictionary types. Previously, every cell in the dictionary array was being added to the distinct values hashset, with this change, we only need to add each value type at most once (never if it is not present) to the array.

# What changes are included in this PR?
A new accumulator (`CountDistinctDictAccumulator`) that is returned by `DistinctCount` in the case that a dictionary array is being counted. There is a fair amount of shared logic between the accumulators so that was also pulled out into helper funcs.

# Are these changes tested?
Added some new unit tests.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (waynexia) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:269]:
```suggestion
    #[cfg(test)]
    fn state(&self) -> Result<Vec<ScalarValue>> {
```

It seems that this is only utilized in tests. Would it be better to addd the cfg annotation?

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (waynexia) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:303]: Are we guaranteed that the value set is consistent among every dictionary array? If not I'm afraid we still need some hash in this accumulator.

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (waynexia) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:803]: what about changing the second dictionary to (reorder the value set)
- value: ["c", "b", "a"]
- keys: [2, 1, None]

**Review by waynexia — COMMENTED** [review]
This is a good improvement :+1: But I have some questions about its correctness when the input dictionaries are not that normalized:

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (jaylmiller) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:269]: This doesn't compile since `state` is part of `Accumulator` trait.

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (jaylmiller) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:303]: I was under the impression that this is the case with how the datafusion accumulators work. I'm fairly new to learning datafusion internals though so I might be wrong about that! 

I'll look into this more that is a good point and a large assumption on my end 😅

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (jaylmiller) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:803]: Yes this would break under my current assumption, will change some stuff and get this working again.

Comment (jaylmiller) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:803]: This is a good test case, thanks 💪

**Comment by jaylmiller** [comment]
@waynexia thanks for correcting my assumption about how normalized dicts 😀. I've made changes correcting this. There is a bit of hashing required but significantly less than before since we only need to hash once for each value type (instead of hashing every cell). 

Also since there was now alot of shared logic between the 2 accumulators, i've pulled that out into funcs so both accumulators can use it.

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (waynexia) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:275]: Sorry I didn't notice this is inside the trait impl block...

**Comment by comphead** [comment]
@jaylmiller thanks for the PR. Would be great to get some knowloedge how the much performance increased?

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (comphead) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:291]: do you think if its possible to merge 2 loops into 1

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:286]:
```suggestion
        let mut seen_map = vec![(false; nvalues];
```

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:753]: 👍 good call to use a different dictionary

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:279]: As an alternate construction, since it is only `update_batch` that varies by the  dictionary key type, I suspect you could make this PR quite a bit smaller by dispatching at runtime to an appropriate type of update_batch. However, that would require a dispatch on each batch, where this PR only requires a single dispatch during planning (at the expense of larger code)

**Review by alamb — APPROVED** [review]
Thank you @jaylmiller  -- this looks great. Thank you @waynexia  for the initial review and ensuring that the dictionaries are handled correctle.

I double checked at the logic now appears to handle different dictionaries correctly -- could you give it one more review @waynexia ?

Also, I wonder if you have had a chance to do any sort of benchmarking to show the improvement?

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:132]: It would be preferable to fix this prior to merge, but I think it would also be fine to merge as is

**Comment by waynexia** [comment]
Sorry for the delay, I plan to review it tomorrow!

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (jaylmiller) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:35]: I was a bit confused about the purpose of `DistinctScalarValues` as well to be honest but I kind of figured it was there for a good reason so I left it in 😅

In terms of the added `ValueSet` alias, I personally thought it made the code a bit more readable but that is kind of subjective of course.

Comment (jaylmiller) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:35]: I think maybe we remove `DistinctScalarValues` alias but keep `ValueSet`?

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (jaylmiller) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:280]: I sortof assumed that `update_batch` is called frequently enough that the extra code was worth but someone more experienced with datafusion would know better than me about this

**Comment by jaylmiller** [comment]
> Also, I wonder if you have had a chance to do any sort of benchmarking to show the improvement?

Currently looking into this. Will post findings today

**Comment by jaylmiller** [comment]
> I think some case in clickbench will be improved.

Ok I'll look into getting some results on these cases

**Comment by jaylmiller** [comment]
ClickBench count distinct query when using dictionary columns is getting killed (this is on main) 🤔

```
❯ CREATE EXTERNAL TABLE hits_base
STORED AS PARQUET
LOCATION 'hits.parquet';
0 rows in set. Query took 0.041 seconds.
❯ CREATE TABLE hits as
select
  arrow_cast("UserID", 'Dictionary(Int32, Utf8)') as "UserID"
FROM hits_base;

0 rows in set. Query took 13.887 seconds.
❯ SELECT COUNT(DISTINCT "UserID") from hits;
Killed
```

"UserID" table is pretty high cardinality though: is there a better clickbench query/column pair to bench with?

**Comment by jackwener** [comment]
cc @sundy-li

**Comment by sundy-li** [comment]
For numeric/string args in distinct, I think it's better to have special states rather than putting the enum into the HashSet.

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (waynexia) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:131]: nit: I would prefer to add the concrete type in the error message

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (waynexia) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:151]:
```suggestion
/// calculating the size of values hashset for fixed length values,
/// taking first batch size * number of batches.
/// This method is faster than full_size(), however it is not suitable for variable length
/// values like strings or complex types
```

style: prefer to use document comments

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (waynexia) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:160]:
```suggestion
}

// calculates the size as accurate as possible, call to this method is expensive
```

style: add empty line between two fns

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (waynexia) [inline datafusion/physical-expr/src/aggregate/count_distinct.rs:261]:
```suggestion
}

impl<K: ArrowDictionaryKeyType + std::marker::Send + std::marker::Sync>
```

style: add empty line between two blocks

**Review by waynexia — COMMENTED** [review]
The current implementation generally looks good to me. I left some little suggestions about style. Looking forward to the bench result 🚀

**Comment by jaylmiller** [comment]
I'm not seeing any noticeable ClickBench improvements when changing "RegionID" column to use dict encoding (setting other columns to use dict encoding cause my datafusion process to be killed...) 

Maybe this change is not worth merging if ClickBench improvements aren't being seen

**Comment by mingmwang** [comment]
@jaylmiller 
How do you do the benchmark test ? I do not get a chance to take a closer look at this PR. 
There is one logical optimization rule `RewriteDistinctAggregate` in DataFusion which will rewrite the single distinct aggregate to normal group by aggregate and replace the distinct aggregator to normal count aggregator.

**Comment by jaylmiller** [comment]
I am essentially just running clickbench queries against my PR and against main. The only change is I am setting RegionID column to a dict:
```
CREATE EXTERNAL TABLE hits_base
STORED AS PARQUET
LOCATION 'hits.parquet';
CREATE TABLE hits as
select 
  arrow_cast("RegionID", 'Dictionary(Int32, Utf8)') as "RegionID"
....
```

Do you have any recommendations, @mingmwang ?

**Comment by jaylmiller** [comment]
Thanks I'll try that out

**Comment by alamb** [comment]
> ClickBench count distinct query when using dictionary columns is getting killed (this is on main as well as the PR) 🤔

I wonder if we can try a smaller subset 🤔 

```sql
❯ CREATE TABLE hits as select
  arrow_cast("UserID", 'Dictionary(Int32, Utf8)') as "UserID"
FROM 'hits.parquet'
limit 10000000;
0 rows in set. Query took 0.776 seconds.
❯ select count(distinct "UserID") from hits;
+-----------------------------+
| COUNT(DISTINCT hits.UserID) |
+-----------------------------+
| 1530334                     |
+-----------------------------+
1 row in set. Query took 71.388 seconds.
```

I will try this on my benchmark machine

**Comment by mingmwang** [comment]
Any luck? 
I think the rule `SingleDistinctToGroupBy` has conflicts with the optimization and rewriting `Distinct` to `Group By` is not always beneficial.  Maybe we should add a configuration to turn on/off this rewriting.

File: datafusion/physical-expr/src/aggregate/count_distinct.rs
Comment (jaylmiller) [reply datafusion/physical-expr/src/aggregate/count_distinct.rs:287]: Thanks for suggestion. Just to clarify, you mean using arrow bitmap, correct? Something like

```rs
let mut bitmap = arrow::array::BooleanBufferBuilder::new(nvalues);
```

**Comment by jaylmiller** [comment]
@mingmwang Sorry for delay. I haven't had a chance to get back to this PR yet (currently working on #5292).

**Comment by alamb** [comment]
Marking as draft to signify this PR has feedback and is not waiting for another review at the moment.

**Comment by alamb** [comment]
Since this has been open for more than a year, closing it down. Feel free to reopen if/when you keep working on it.

**Pull request closed without merging**
